### ISSUE TITLE
DefaultResourceLockCoordinationService implementation may cause Gradle hang

### ISSUE THREAD

**Issue by blindpirate** [gradle/gradle#6696]
### Context

Recently one of our builds hangs at:

```
 "Daemon worker" #18 prio=5 os_prio=0 tid=0x0000000059280000 nid=0x1878 in Object.wait() [0x000000005b3ac000]
 java.lang.Thread.State: BLOCKED (on object monitor)
 	at java.lang.Object.wait(Native Method)
 	- waiting on <0x0000000080092300> (a java.lang.Object)
 	at java.lang.Object.wait(Object.java:502)
 	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:51)
 	- locked <0x0000000080092300> (a java.lang.Object)
 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:161)
...

 "Task worker for ':' Thread 3" #146 prio=5 os_prio=0 tid=0x0000000057556000 nid=0x2204 in Object.wait() [0x000000005f3ce000]
 java.lang.Thread.State: BLOCKED (on object monitor)
 	at java.lang.Object.wait(Native Method)
 	- waiting on <0x0000000080092300> (a java.lang.Object)
 	at java.lang.Object.wait(Object.java:502)
 	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:51)
 	- locked <0x0000000080092300> (a java.lang.Object)
 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:161)
 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
 	at java.lang.Thread.run(Thread.java:748)
 
 "Task worker for ':'" #144 prio=5 os_prio=0 tid=0x0000000057554800 nid=0x2b80 in Object.wait() [0x000000005f9ce000]
 java.lang.Thread.State: BLOCKED (on object monitor)
 	at java.lang.Object.wait(Native Method)
 	- waiting on <0x0000000080092300> (a java.lang.Object)
 	at java.lang.Object.wait(Object.java:502)
 	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:51)
 	- locked <0x0000000080092300> (a java.lang.Object)
 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.executeWithWork(DefaultTaskPlanExecutor.java:161)
 	at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$ExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
 	at java.lang.Thread.run(Thread.java:748)
```

In Gradle implementation, if `maxParallelForks` is set to 4, there will be 4 threads available for task execution:

- Daemon worker
- Task worker for ':'
- Task worker for ':' Thread 2
- Task worker for ':' Thread 3

We can see `Daemon worker`/`Task worker for ':'`/`Task worker for ':' Thread 3` are all stuck at `DefaultResourceLockCoordinationService.lock.wait()`, but `Task worker for ':' Thread 2` thread is missing. [`DefaultResourceLockCoordinationService.withStateLock`](https://github.com/gradle/gradle/blob/c1ebd6513e62573713831842c04d1c5e4307d8c6/subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java#L38) method has only 3 exits:

- `FINISHED` return true (impossible because other threads don't get notification from `Object.wait`)
- `FAILED` return false (impossible because [this action](https://github.com/gradle/gradle/blob/0f25cd25d84c49c9b82a9d897a47ecb5c74fb714/subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskPlanExecutor.java#L161) can only return `FINISHED` and `RETRY`)
- throw exception

When an exception is thrown, [this block](https://github.com/gradle/gradle/blob/c1ebd6513e62573713831842c04d1c5e4307d8c6/subprojects/base-services/src/main/java/org/gradle/internal/resources/DefaultResourceLockCoordinationService.java#L66) doesn't invoke `lock.notifyAll()`, which under certain conditions may cause the whole process hang.

[This unit test](https://github.com/gradle/gradle/commit/53d7876097ae8f0c5c0166c512ad2da801532f65) can reproduce this issue.

@ghale can you please confirm this issue?

[Test_Coverage_-_Platform_Java8_Oracle_Windows_Test_Coverage_-_Platform_Java8_Oracle_Windows_modelGroovy_9946.txt](https://github.com/gradle/gradle/files/2364992/Test_Coverage_-_Platform_Java8_Oracle_Windows_Test_Coverage_-_Platform_Java8_Oracle_Windows_modelGroovy_9946.txt)

**Comment by oehme** [comment]
That test is invalid - No locks are being acquired or released, it's just an endless loop. This test would even fail without the exception and just a single thread returning `RETRY` once.

The catch block is correct, it shouldn't notify because it didn't actually acquire any locks. 

The problem is probably some other thread that acquired a worker lease and never returned it.

**Comment by blindpirate** [comment]
I think it is an endless loop with exits - all threads should eventually exit the loop. If the test has only one thread and simply returns `RETRY`, it doesn't make any sense because it can't exit the `DefaultResourceLockCoordinationService.withStateLock` loop at all.

> The catch block is correct, it shouldn't notify because it didn't actually acquire any locks.

If it exits via catch block and other threads happen to be waiting for `lock.wait()`, they won't get any notifications and will wait forever, is this expected behaviour? Am I missing something?

According to the thread dump, `Task worker for ':' Thread 2` thread is missing and the only possible exit is catch block, so I think the catch block is the problem. Can a thread die from some other reason?

**Comment by asockalingam** [comment]
what is the resolution for this? how do we avoid this?

**Comment by mcumings** [comment]
There definitely seems to be a major issue here.  We have a vertically scaled machine that should be able to chew through builds (80 CPU threads on NVMe) but it is slower than building on MacBook Pros filled with corpware.  Our build does seem to progress, though at an absolute crawl.  FWIW, it appears as though some threads eventually unblock for a task or two, then go back into the blocked state.

I've attached a stack dump taken while this build was running.

[gradlestack.txt](https://github.com/gradle/gradle/files/5983507/gradlestack.txt)

**Comment by ghale** [comment]
Hey @mcumings I think what you are describing is a different issue.  In your case it sounds like it doesn't hang, but it performs poorly at a very high degree of parallelism/number of tasks.  Basically, you're ending up with more contention than actual work being done.  I would suggest that you open a separate issue, and include some flame graphs using   https://github.com/gradle/gradle-profiler (use the async-profiler option).

**Comment by mcumings** [comment]
Roger that, @ghale.  I'll double back on this when I get a few cycles and see if I can spin this up.

**Comment by octylFractal** [comment]
Closing as obsolete.
